### PR TITLE
 Re-Organize the pkutil library

### PR DESCRIPTION
**Pull request by mourya007**
https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil has some overlap with the other libraries. It is also kubeadm-specific but has several dependencies in client-go that nobody else uses.

#71004

None

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
> https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil has some overlap with the other libraries. It is also kubeadm-specific but has several dependencies in client-go that nobody else uses.

that is true but we seem to be moving things back and forth.
these particular helpers make sense for a wider range of consumers too.

the original plan was to have such utils in a new package outside of kubeadm.
so maybe not in client-go but somewhere else.

/assign @liztio @timothysc @luxas @awly

**Comment by neolit123** [comment]
> This seems reasonable, somewhere under k8s.io/kubernetes/pkg/util perhaps?

SGTM, would love to see feedback from @liztio and @luxas on this.

**Comment by mourya007** [comment]
> @mourya007 please also remove `NewSignedCert` from `staging/src/k8s.io/client-go/util/cert/cert.go`

Done

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go
Comment (awly) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: `certutil.Config`

Comment (mourya007) [reply cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: Hey, you want this to change S/cfg config/certutil.Config ?

Comment (awly) [reply cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: Yeah, otherwise this should fail to compile

Comment (mourya007) [reply cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: Done, Please check

**Comment by awly** [comment]
Also remove `rsaKeySize` in staging/src/k8s.io/client-go/util/cert/cert.go
LGTM otherwise, but kubeadm owners should have the final say.

**Comment by mourya007** [comment]
> Also remove `rsaKeySize` in staging/src/k8s.io/client-go/util/cert/cert.go
> LGTM otherwise, but kubeadm owners should have the final say.

Done changes, Please have a look.

**Comment by mikedanese** [comment]
Thanks!

/lgtm
/approve

**Comment by timothysc** [comment]
/ok-to-test 
/hold 
/assign @timothysc @liztio 

@liztio could you please verify and unhold if ok.

**Review by liztio — APPROVED** [review]
Makes sense to me, thanks for the help!

**Comment by awly** [comment]
@mourya007 the build seems to fail:

```
I0125 17:45:16.968] /bazel-scratch/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/sandbox/linux-sandbox/2329/execroot/__main__/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:192:4: undefined: cert.EncodeCertPEM
I0125 17:45:16.968] /bazel-scratch/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/sandbox/linux-sandbox/2329/execroot/__main__/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:213:3: undefined: cert.EncodeCertPEM
I0125 17:45:16.968] /bazel-scratch/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/sandbox/linux-sandbox/2329/execroot/__main__/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:215:3: undefined: cert.EncodeCertPEM
```

Please make sure `make bazel-build` passes locally.

**Comment by mourya007** [comment]
> @mourya007 the build seems to fail:
> 
> ```
> I0125 17:45:16.968] /bazel-scratch/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/sandbox/linux-sandbox/2329/execroot/__main__/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:192:4: undefined: cert.EncodeCertPEM
> I0125 17:45:16.968] /bazel-scratch/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/sandbox/linux-sandbox/2329/execroot/__main__/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:213:3: undefined: cert.EncodeCertPEM
> I0125 17:45:16.968] /bazel-scratch/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/sandbox/linux-sandbox/2329/execroot/__main__/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:215:3: undefined: cert.EncodeCertPEM
> ```
> 
> Please make sure `make bazel-build` passes locally.

@awly yes, I am working on it.

**Comment by mourya007** [comment]
Thank you for the review

**Comment by neolit123** [comment]
@mourya007 `pull-kubernetes-node-e2e` is broken today. try `/retest` tomorrow - hopefully will be fixed by then. thanks.

**Comment by mourya007** [comment]
> @mourya007 `pull-kubernetes-node-e2e` is broken today. try `/retest` tomorrow - hopefully will be fixed by then. thanks.

Ok, Sure I will do the same. Thanks for the help

**Comment by lavalamp** [comment]
@mourya007 @mikedanese This change broke the public client interface without leaving any breadcrumbs or release note. It's probably OK to move this stuff but we need to be deliberate about it.

@liggitt How's that public interface contract checker going? :) :) :)

**Comment by mikedanese** [comment]
Thanks, we were aware of this but thought this was within the compatibility guidelines for client-go since this was introduced at a major version. In the future, how should we handle these type of cleanups? We have some utility packages in client-go that still need to be cleaned up or moved out.

**Comment by lavalamp** [comment]
Yeah, I'm not saying it has to stay. Jus that there still needs to be a
release note, or an entry in the changelog, or an announcement email, or
leave a breadcrumb in the code for a release or something.

**Comment by mikedanese** [comment]
Ok, we'll keep that in mind. Thanks!

**Comment by liggitt** [comment]
> @liggitt How's that public interface contract checker going? :) :) :)

https://github.com/liggitt/kubernetes/commits/apidiff-master

```
$ hack/verify-apidiff.sh 

k8s.io/apimachinery ==========

v1.14.0 =====
pkg.api.meta:
- List.GetRemainingItemCount: added
- List.SetRemainingItemCount: added
pkg.api.resource:
- (*Quantity).Copy: removed
pkg.apis.meta.internalversion:
- Codecs: removed
- ParameterCodec: removed
pkg.apis.meta.v1:
- (*ObjectMeta).GetInitializers: removed
- (*ObjectMeta).SetInitializers: removed
- Fields: removed
- Initializer: removed
- Initializers: removed
- ListInterface.GetRemainingItemCount: added
- ListInterface.SetRemainingItemCount: added
- ManagedFieldsEntry.Fields: removed
- Object.GetInitializers: removed
- Object.SetInitializers: removed
- ObjectMeta.Initializers: removed
pkg.apis.meta.v1.unstructured:
- (*Unstructured).GetInitializers: removed
- (*Unstructured).SetInitializers: removed
- JSONFallbackEncoder: removed
pkg.apis.meta.v1beta1:
- PartialObjectMetadataList.Items: changed from []*PartialObjectMetadata to []k8s.io/apimachinery/pkg/apis/meta/v1.PartialObjectMetadata
pkg.apis.testapigroup:
- (*ObjectMeta).GetInitializers, method set of *Carp: removed
- (*ObjectMeta).SetInitializers, method set of *Carp: removed
- Carp.Initializers: removed
pkg.apis.testapigroup.v1:
- (*ObjectMeta).GetInitializers, method set of *Carp: removed
- (*ObjectMeta).SetInitializers, method set of *Carp: removed
- Carp.Initializers: removed
pkg.runtime:
- Codec.Identifier: added
- Encoder.Identifier: added
- GroupVersioner.Identifier: added
- GroupVersioners: removed
- Serializer.Identifier: added
- Unstructured.NewEmptyInstance: added
pkg.runtime.serializer:
- DirectCodecFactory: removed
- NewCodecFactory: changed from func(*k8s.io/apimachinery/pkg/runtime.Scheme) CodecFactory to func(*k8s.io/apimachinery/pkg/runtime.Scheme, ...CodecFactoryOptionsMutator) CodecFactory
pkg.runtime.serializer.protobuf:
- NewRawSerializer: changed from func(k8s.io/apimachinery/pkg/runtime.ObjectCreater, k8s.io/apimachinery/pkg/runtime.ObjectTyper, string) *RawSerializer to func(k8s.io/apimachinery/pkg/runtime.ObjectCreater, k8s.io/apimachinery/pkg/runtime.ObjectTyper) *RawSerializer
- NewSerializer: changed from func(k8s.io/apimachinery/pkg/runtime.ObjectCreater, k8s.io/apimachinery/pkg/runtime.ObjectTyper, string) *Serializer to func(k8s.io/apimachinery/pkg/runtime.ObjectCreater, k8s.io/apimachinery/pkg/runtime.ObjectTyper) *Serializer
pkg.runtime.serializer.versioning:
- DirectDecoder: removed
- DirectEncoder: removed
pkg.util.sets:
- Byte.Delete: changed from func(...byte) to func(...byte) Byte
- Byte.Insert: changed from func(...byte) to func(...byte) Byte
- Int.Delete: changed from func(...int) to func(...int) Int
- Int.Insert: changed from func(...int) to func(...int) Int
- Int64.Delete: changed from func(...int64) to func(...int64) Int64
- Int64.Insert: changed from func(...int64) to func(...int64) Int64
- String.Delete: changed from func(...string) to func(...string) String
- String.Insert: changed from func(...string) to func(...string) String
pkg.watch:
- NewStreamWatcher: changed from func(Decoder) *StreamWatcher to func(Decoder, Reporter) *StreamWatcher

v1.15.0 =====
pkg.api.resource:
- (*Quantity).Copy: removed
pkg.apis.meta.internalversion:
- Codecs: removed
- ParameterCodec: removed
pkg.apis.meta.v1:
- (*ObjectMeta).GetInitializers, method set of *PartialObjectMetadata: removed
- (*ObjectMeta).GetInitializers: removed
- (*ObjectMeta).SetInitializers, method set of *PartialObjectMetadata: removed
- (*ObjectMeta).SetInitializers: removed
- Fields: removed
- Initializer: removed
- Initializers: removed
- ManagedFieldsEntry.Fields: removed
- Object.GetInitializers: removed
- Object.SetInitializers: removed
- ObjectMeta.Initializers: removed
- PartialObjectMetadata.Initializers: removed
pkg.apis.meta.v1.unstructured:
- (*Unstructured).GetInitializers: removed
- (*Unstructured).SetInitializers: removed
- JSONFallbackEncoder: removed
pkg.apis.testapigroup:
- (*ObjectMeta).GetInitializers, method set of *Carp: removed
- (*ObjectMeta).SetInitializers, method set of *Carp: removed
- Carp.Initializers: removed
pkg.apis.testapigroup.v1:
- (*ObjectMeta).GetInitializers, method set of *Carp: removed
- (*ObjectMeta).SetInitializers, method set of *Carp: removed
- Carp.Initializers: removed
pkg.runtime:
- Codec.Identifier: added
- Encoder.Identifier: added
- GroupVersioner.Identifier: added
- GroupVersioners: removed
- Serializer.Identifier: added
pkg.runtime.serializer:
- DirectCodecFactory: removed
- NewCodecFactory: changed from func(*k8s.io/apimachinery/pkg/runtime.Scheme) CodecFactory to func(*k8s.io/apimachinery/pkg/runtime.Scheme, ...CodecFactoryOptionsMutator) CodecFactory
pkg.runtime.serializer.versioning:
- DirectDecoder: removed
- DirectEncoder: removed
pkg.util.sets:
- Byte.Delete: changed from func(...byte) to func(...byte) Byte
- Byte.Insert: changed from func(...byte) to func(...byte) Byte
- Int.Delete: changed from func(...int) to func(...int) Int
- Int.Insert: changed from func(...int) to func(...int) Int
- Int32.Delete: changed from func(...int32) to func(...int32) Int32
- Int32.Insert: changed from func(...int32) to func(...int32) Int32
- Int64.Delete: changed from func(...int64) to func(...int64) Int64
- Int64.Insert: changed from func(...int64) to func(...int64) Int64
- String.Delete: changed from func(...string) to func(...string) String
- String.Insert: changed from func(...string) to func(...string) String

v1.16.0 =====
pkg.apis.meta.internalversion:
- Codecs: removed
- ParameterCodec: removed
pkg.apis.meta.v1.unstructured:
- JSONFallbackEncoder: removed
pkg.runtime:
- Codec.Identifier: added
- Encoder.Identifier: added
- GroupVersioner.Identifier: added
- GroupVersioners: removed
- Serializer.Identifier: added
```

**Pull request merged**
